cli/run_util.go

```diff
@@ -540,14 +583,66 @@ func (r WorkflowRunner) activateAndRunSteps(
 	secrets []envmanModels.EnvironmentItemModel,
 	isLastWorkflow bool,
 	tracker analytics.Tracker,
-	workflowIDProperties coreanalytics.Properties) models.BuildRunResultsModel {
+	workflowIDProperties coreanalytics.Properties,
+	workflowID string,
+) models.BuildRunResultsModel {
 	log.Debug("[BITRISE_CLI] - Activating and running steps")
 
 	if len(workflow.Steps) == 0 {
 		log.Warnf("%s workflow has no steps to run, moving on to the next workflow...", workflow.Title)
 		return buildRunResults
 	}
 
+	err := tools.EnvmanInit(configs.InputEnvstorePath, true)
+	if err != nil {
+		log.Debugf("Couldn't initialize envman.")
+	}
+	err = tools.EnvmanAddEnvs(configs.InputEnvstorePath, *environments)
+	if err != nil {
+		log.Debugf("Couldn't add envs.")
+	}
+	envList, err := tools.EnvmanReadEnvList(configs.InputEnvstorePath)
+	if err != nil {
+		log.Debugf("Couldn't read envs from envman.")
+	}
+
+	serviceContainers, err := r.dockerManager.StartServiceContainers(workflow.Services, workflowID, envList)
+	if err != nil {
+		log.Errorf("❌ Some services failed to start properly!")
+	}
+
+	defer func() {
+		for _, container := range serviceContainers {
+			if err := container.Destroy(); err != nil {
+				log.Errorf("Attempted to stop the docker container for service: %s", container.Name)
+			}
+		}
+	}()
+
+	if workflow.Container.Image != "" {
+		log.Infof("ℹ️ Running workflow in docker container: %s", workflow.Container.Image)
+
+		if err := r.dockerManager.Login(workflow.Container, envList); err != nil {
+			log.Errorf("%s workflow has docker credentials provided, but the authentication failed.", workflow.Title)
```

Comment:
I think we should abort in case of this error

Reply:
The original logic was also not returning errors here, instead of letting the steps eventually fail where it is handled properly (little bit further in this file), so the build can finish properly with the correct status.

Reply:
Okay, this convinced me, thank you!